components/brave_wallet/browser/eth_abi_decoder.cc

```diff
@@ -0,0 +1,292 @@
+/* Copyright (c) 2022 The Brave Authors. All rights reserved.
+ * This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this file,
+ * You can obtain one at http://mozilla.org/MPL/2.0/. */
+
+#include "brave/components/brave_wallet/browser/eth_abi_decoder.h"
+
+#include <limits>
+#include <map>
+
+#include "base/strings/strcat.h"
+#include "base/strings/string_piece.h"
+#include "base/strings/string_util.h"
+#include "brave/components/brave_wallet/common/hex_utils.h"
+
+// This file implements decoding rules of calldata according to the Contract
+// Application Binary Interface (ABI).
+//
+// Method arguments are packed in chunks of 32 bytes, with types smaller than
+// 32 bytes zero padded if necessary. Fixed-size types like uint256, address
+// are represented within the available 32 bytes, however, dynamic types like
+// bytes and arrays follow head-tail encoding. In this scheme, the data is
+// packaged at the tail-end of the transaction's calldata. The arguments are
+// references into the calldata where the content is.
+//
+// Parsing of complex cases such as composite types, nested arrays, etc., are
+// not implemented. The decoder only extracts the calldata tail reference in
+// such cases.
+//
+// Function selector (first 4 bytes) should NOT be part of the calldata being
+// parsed.
+//
+// References:
+//   https://docs.soliditylang.org/en/latest/abi-spec.html
+
+namespace brave_wallet {
+
+namespace {
+// GetAddressFromData extracts an Ethereum address from the calldata at the
+// specified index. The address type is static and 32-bytes wide, but we
+// only consider the last 20 bytes, discarding the leading 12 bytes of
+// 0-padded chars.
+//
+// In the future, addresses in Ethereum may become 32 bytes long:
+// https://ethereum-magicians.org/t/increasing-address-size-from-20-to-32-bytes
+//
+// The parsed address value is prefixed by "0x".
+bool GetAddressFromData(const std::string& input,
+                        size_t offset,
+                        std::string& arg) {  // NOLINT(runtime/references)
+  arg = "0x" + input.substr(offset + 24, 40);
+  return true;
+}
+
+// GetSizeFromData extracts a 32-byte wide size_t value from the calldata
+// at the specified offset.
+//
+// Using this function to extract an integer outside the range of size_t is
+// considered an error. Ideal candidates are calldata tail references, length
+// of dynamic types, etc.
+bool GetSizeFromData(const std::string& input,
```

Comment:
Also here(and other methods) instead of string + offset pair it might be much more clear to use StringPiece

Reply:
Can we also have a function that supports hex to uin256 conversion without requiring leading 0x? So having StringPiece in args we just check if it has at least 64 chars in it and pass it conversion function as is.

Reply:
> Also here(and other methods) instead of string + offset pair it might be much more clear to use StringPiece

While working on the above modification, I realised offsets would be a good choice here because we may need to jump around within the calldata in order to parse dynamic types. Dynamic types are actually 32-byte pointers to the calldata tail section. Since these pointers are simply indexes relative to the start of the calldata string, it becomes a problem if we modify the StringPiece as we parse the calldata vs using offsets.

Maybe I'm missing something, so open to your thoughts on this.

Reply:
> Can we also have a function that supports hex to uin256 conversion without requiring leading 0x? So having StringPiece in args we just check if it has at least 64 chars in it and pass it conversion function as is.

I did a small refactoring to make things simpler: 5cc39c626d9a8d3aea3ae6f33c39930301e6399a, to enable more code-reuse. Also note that the reason why we did not just check 64 chars and return it as-is for uint256, was to get rid of extra leading 0s. So we decode the hex into a uint256_t and then back to hex.

Reply:
_string + offset pair vs StringPiece_
let's stay with the first option for start then

_was to get rid of extra leading 0s_
Ok, I understand